libs/providers/unleash-web/src/lib/unleash-web-provider.ts

```diff
@@ -0,0 +1,164 @@
+        message: 'Ready',
+      });
+    });
+    this._client?.on('update', () => {
+      this._logger?.info('Unleash update event received');
+      this.events.emit(ProviderEvents.ConfigurationChanged, {
+        message: 'Flags changed',
+      });
+    });
+    this._client?.on('error', () => {
+      this._logger?.info('Unleash error event received');
+      this.events.emit(ProviderEvents.Error, {
+        message: 'Error',
+      });
+    });
+    this._client?.on('recovered', () => {
+      this._logger?.info('Unleash recovered event received');
+      this.events.emit(ProviderEvents.Ready, {
+        message: 'Recovered',
+      });
+    });
+  }
+
+  async onContextChange(_oldContext: EvaluationContext, newContext: EvaluationContext): Promise<void> {
+    const unleashContext = new Map();
+    const properties = new Map();
+    Object.keys(newContext).forEach((key) => {
+      switch (key) {
+        case 'appName':
+        case 'userId':
+        case 'environment':
+        case 'remoteAddress':
+        case 'sessionId':
+        case 'currentTime':
+          unleashContext.set(key, newContext[key]);
+          break;
+        default:
+          properties.set(key, newContext[key]);
+          break;
+      }
+    });
+    if (properties.size > 0) {
+      unleashContext.set('properties', Object.fromEntries(properties));
+    }
+    await this._client?.updateContext(Object.fromEntries(unleashContext));
+    this._logger?.info('Unleash context updated');
+  }
+
+  async onClose() {
+    this._logger?.info('closing Unleash client');
+    this._client?.stop();
+  }
+
+  resolveBooleanEvaluation(flagKey: string): ResolutionDetails<boolean> {
+    const resp = this._client?.isEnabled(flagKey);
+    if (typeof resp === 'undefined') {
+      throw new FlagNotFoundError();
+    }
+    return {
+      value: resp,
+    };
+  }
+
+  resolveStringEvaluation(flagKey: string, defaultValue: string): ResolutionDetails<string> {
+    return this.evaluate(flagKey, defaultValue);
+  }
+
+  resolveNumberEvaluation(flagKey: string, defaultValue: number): ResolutionDetails<number> {
+    const resolutionDetails = this.evaluate(flagKey, defaultValue);
+    resolutionDetails.value = Number(resolutionDetails.value);
+    return resolutionDetails;
+  }
+
+  resolveObjectEvaluation<U extends JsonValue>(flagKey: string, defaultValue: U): ResolutionDetails<U> {
+    return this.evaluate(flagKey, defaultValue);
+  }
+
+  private evaluate<T>(flagKey: string, defaultValue: T): ResolutionDetails<T> {
+    const evaluatedVariant = this._client?.getVariant(flagKey);
+    let value;
+    let variant;
+    if (typeof evaluatedVariant === 'undefined') {
+      throw new FlagNotFoundError();
+    }
+
+    if (evaluatedVariant.name === 'disabled') {
+      value = defaultValue as T;
+    } else {
+      variant = evaluatedVariant.name;
+      value = evaluatedVariant.payload?.value;
+    }
+    return {
+      variant: variant,
+      value: value as T,
```

Comment:
Please add a check if the type is convertible to the type `T`, you can do it using the default value. 

E.g. for numbers, add something like: 
```ts
if (typeof defaultValue === Number && isNaN(parseFloat(value))) {
  throw new TypeMismatchError()
} else if (typeof defaultValue ===
.....
```

And then instead of doing the conversion in the specific methods you can do them here to ensure type safety at a central place.

Reply:
yes makes sense, I will look into this tomorrow.